File: Java-Kotlin/src/main/java/array/_34.java
package array;

public class _34 {
    public int[] searchRange(int[] nums, int target) {
        if (nums == null || nums.length <= 0) {
            return new int[] {-1, -1};
        }

        int oneKnownIndex = getKnownIndex(nums, target);
        if (oneKnownIndex == -1) {
            return new int[] {-1, -1};
        }

        int left = oneKnownIndex;
        while (left > 0 && nums[left - 1] == nums[left]) {
            left--;
        }
        int right = oneKnownIndex;
        while (right < nums.length - 1 && nums[right] == nums[right + 1]) {
            right++;
        }
        return new int[] {left, right};
    }

    private int getKnownIndex(int[] nums, int target) {
        int start = 0;
        int end = nums.length - 1;
        while (start <= end) {
            int mid = (start + end) / 2;
            if (nums[mid] == target) {
                return mid;
            } else if (nums[mid] < target) {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        return -1;
    }
}

File: Go-Solutions/34.go
package main

// 套模板就可以了
func searchRange(nums []int, target int) []int {
	left := getFarLeftIndex(nums, target)
	right := getFarRightIndex(nums, target)
	return []int{left, right}
}

func getFarRightIndex(nums []int, target int) int {
	left, right := 0, len(nums) - 1
	for left <= right {
		mid := left + (right - left)>> 1
		if nums[mid] > target {
			right = mid - 1
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			if (mid == len(nums) - 1) || (nums[mid + 1] != target) {
				return mid
			}
			left = mid + 1
		}
	}
	return -1
}

func getFarLeftIndex(nums []int, target int) int {
	left, right := 0, len(nums) - 1
	for left <= right {
		mid := left + (right - left) >> 1
		if nums[mid] > target {
			right = mid - 1
		} else if nums[mid] < target {
			left = mid + 1
		} else {
			if (mid == 0) || (nums[mid-1] != target) {
				return mid
			}
			right = mid - 1
		}
	}
	return -1
}
